### PR TITLE
dockerfile based binary building

### PR DESCRIPTION
**Pull request by tonistiigi**
Using cross-compilation toolchains that work from any platform
Adds darwin/arm64 support and bake targets. Static and dynamic
binary targets are available, both with glibc and musl.

### Build static binary for the current architecture.

```
VERSION=20.10.6 docker buildx bake
```

Equivalent:
```
docker buildx build --platform=local --output ./build
```

Out of tree:
```
docker buildx bake "git://github.com/tonistiigi/docker-cli#xx-build"

docker buildx bake "git://github.com/tonistiigi/docker-cli#xx-build" binary
```

### Any other platform

```
docker buildx bake --set binary.platform=darwin/arm64
```

```
docker buildx build --platform=darwin/arm64 --output ./build
```

### Build dynamic binaries

```
docker buildx bake dynbinary
USE_GLIBC=1 docker buildx bake dynbinary # debian based build
```

### All supported platforms

```
docker buildx bake cross
docker buildx bake dynbinary-cross
```

Here is the output: https://gist.github.com/tonistiigi/7f16d059ae97d790bdde314f23f2c94d

Toolchain build from https://github.com/tonistiigi/xx/pull/14 (some early docs: https://gist.github.com/tonistiigi/ebcac239dd3baef6747fc54f3f73cbba)

TODO:
- Windows resources with windres not currently added
- Update the callers
- Remove more old targets
- Examples

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

**Comment by thaJeztah** [comment]
I guess this means we can remove (a couple of) the dockerfiles in https://github.com/docker/cli/tree/master/dockerfiles, correct?

**Comment by tonistiigi** [comment]
@thaJeztah Yes, mentioned in https://github.com/docker/cli/pull/2993/commits/61ae3bb7b5221dd39878665f1b1dd42a6ccfd1cf . Only did the parts that were not optional to the binary building.

**Review by thaJeztah — COMMENTED** [review]
(apparently I had some review comments, in draft but not submitted; let me post them 😂)

File: scripts/build/binary [hunk 1]
```diff
@@ -1,14 +1,80 @@
-#!/usr/bin/env bash
+#!/usr/bin/env sh
 #
 # Build a static binary for the host OS/ARCH
```
Comment (tiborvass) [inline scripts/build/binary:3]: remove comment

**Review by tiborvass — APPROVED** [review]
We can revert some of the Makefile changes for 20.10 branch to make less big changes there.

**Review by thaJeztah — APPROVED** [review]
lgtm

File: scripts/build/.variables [hunk 2]
```diff
@@ -1,8 +1,10 @@
-#!/usr/bin/env bash
+#!/usr/bin/env sh
 set -eu
 
+TARGET=${TARGET:-"build"}
+
 PLATFORM=${PLATFORM:-}
-VERSION=${VERSION:-"unknown-version"}
+VERSION=${VERSION:-$(git describe --match 'v[0-9]*' --dirty='.m' --always --tags | sed 's/^v//' 2>/dev/null || echo "unknown-version" )}
```
[the remark this answers is not in the document]
Comment (tianon) [reply scripts/build/.variables:7]: Ah, and worse, the `||` only takes effect if the `sed` fails, so I'd suggest something like this instead:

```suggestion
VERSION=${VERSION:-$(git describe --match 'v[0-9]*' --dirty='.m' --always --tags 2>/dev/null || echo "unknown-version" | sed 's/^v//' )}
```

**Pull request merged**